Review Threads Operating Successfully

aclowmclaughlin opened the pull request:
Initial RTOS thread setup- all the threads are at least templated out. Model thread is running.

aclowmclaughlin committed:
Initial Commit

aclowmclaughlin committed:
Created Simulink Thread and PowertrainCAN Receive thread. Had to modify EVT-core, gonna need to make a branch there for stuff to actually work. CMake is having issues finding tx_api.h, which I thought I had resolved but apparently it is not happy about it still.

aclowmclaughlin committed:
random renaming fixes

aclowmclaughlin committed:
Updated MCuC target to use RTOS.

aclowmclaughlin committed:
Stubbed out all Threads.

aclowmclaughlin committed:
Stubbed out all Threads.

aclowmclaughlin committed:
Threads templating all done, needs CANopen for the Accessory CAN Thread to do anything.

aclowmclaughlin committed:
Switched to different branch of EVT-core

aclowmclaughlin committed:
Updated Github Actions

aclowmclaughlin committed:
Added PowerTrain Can Message IDs

aclowmclaughlin committed:
whoopsies

aclowmclaughlin committed:
whoopsies part 2.

aclowmclaughlin committed:
random commit.

aclowmclaughlin committed:
just need the build test to run again.

aclowmclaughlin committed:
updated evtcore

aclowmclaughlin committed:
updates so that MCuC can use AccessoryCAN

aclowmclaughlin committed:
More Updates for Accessory CAN

aclowmclaughlin committed:
Fixed MCuC Object Dictionary.

aclowmclaughlin committed:
updated MCuC_Model to new version.

aclowmclaughlin committed:
set MCuC values to default false for safety.

aclowmclaughlin committed:
Changes for Imagine stuff.

aclowmclaughlin committed:
Changed back stuff that was messed up by imagine, fixed MCuC Accessory CAN

aclowmclaughlin committed:
updated to right evt-core branch version

Commit:
Applied Formatting Changes During GitHub Build

DiegoLHendrix commented on include/MCuC.hpp:
```diff
@@ -10,97 +10,95 @@
 #include <core/utils/types/FixedQueue.hpp>
```
Remove this ```#include``` if you do not plan on using it.

DiegoLHendrix commented on include/MCuC.hpp:
```diff
@@ -10,97 +10,95 @@
 #include <core/utils/types/FixedQueue.hpp>
 #include <models/MCuC_Model.hpp>
 
+#include <core/rtos/Initializable.hpp>
```
Correct me if I am wrong but shouldn't there be an explanation for what the MCuC does at the top of this file?

DiegoLHendrix commented on include/PowertrainCAN.hpp:
```diff
@@ -3,41 +3,60 @@
 #include <core/utils/types/FixedQueue.hpp>
 #include <cstddef>
 #include <cstdint>
 #include <cstring>
```
Remove unused libraries here as well.

DiegoLHendrix commented on include/PowertrainCAN.hpp:
```diff
@@ -3,41 +3,60 @@
+     * An enumeration of the Powertrain CAN message ids that are relevant to the MCuC and Hardmon.
      */
     enum MessageIDs {
+        // motor controller recieve message ids
```
```suggestion
        // motor controller receive message ids
```

DiegoLHendrix commented on targets/REV3-Hardmon/main.cpp:
```diff
@@ -37,87 +132,49 @@ namespace time = core::time;
 
 /**
  * Interrupt handler to get CAN messages from the powertrain CAN line.
  * @param message[in] the passed in in CAN message that was read.
```
```suggestion
 * @param message[in] the passed in CAN message that was read.
```

DiegoLHendrix commented on targets/REV3-Hardmon/main.cpp:
```diff
@@ -4,24 +4,119 @@
+ */
+typedef struct {
+    vcu::Hardmon* hardmon;
+    //todo: once threadsafe canopen is implemented, this should take an instance of that.
```
This is might just me but every other file with a TODO is in all caps so just for consistency keep the same format here. 
```suggestion
    //TODO: once threadsafe canopen is implemented, this should take an instance of that.
```

DiegoLHendrix commented on targets/REV3-Hardmon/main.cpp:
```diff
@@ -151,13 +208,205 @@ int main() {
+    rtos::TXError error;
+    while (true) {
+        log::LOGGER.log(core::log::Logger::LogLevel::DEBUG, "Health Thread Triggered.");
+        //do healththread stuff
```
I'm just nitpicking at this point ngl 💀.
```suggestion
        //do healthThread stuff
```

DiegoLHendrix commented on targets/REV3-MCUC/main.cpp:
```diff
@@ -50,8 +146,8 @@ void accessoryCANOpenInterrupt(IO::CANMessage& message, void* priv) {
  * @param message[in] the passed in in CAN message that was read.
```
```suggestion
 * @param message[in] the passed in CAN message that was read.
```

DiegoLHendrix commented on targets/REV3-MCUC/main.cpp:
```diff
@@ -50,8 +146,8 @@ void accessoryCANOpenInterrupt(IO::CANMessage& message, void* priv) {
  * @param message[in] the passed in in CAN message that was read.
  * @param priv[in] the private data this mesasge requires. Should be the mcuc instance we want to update.
```
```suggestion
 * @param priv[in] the private data this message requires. Should be the mcuc instance we want to update.
```

DiegoLHendrix commented on targets/REV3-Hardmon/main.cpp:
```diff
@@ -151,13 +208,205 @@ int main() {
+[[noreturn]] void accessoryCanReceiveThreadEntry(accessoryCanReceiveThreadArgs_t* args) {
+    log::LOGGER.log(core::log::Logger::LogLevel::DEBUG, "Accessory Can Thread Started.");
+    rtos::TXError error;
+    //    while(true) {
```
I think I am missing something. Why is this while loop commented out?

aclowmclaughlin replied:
The Hardmon dictionary is super incomplete (I was using the MCuC one to figure out what we needed) so running it will probably result in a crash, or at least undefined behaviour. Added a todo so it's more obvious what's wrong.

mjh9585 replied:
Is it safe to let a thread just exit like this without calling a thread delete or something? I would at least keep the loop and just put a suspend or long sleep in it to ensure it doesn't cause a problem.

aclowmclaughlin replied:
It's generally safe, I believe that when a thread reaches the end of it's entry function it is set to sleep automatically. Technically it's likely not the best solution, but this won't be how it works in the actual code (we definitely need the accessory can thread), so it will be fine for now.

mjh9585 replied:
Yeah, but I was mostly thinking of leaving the loop and just sleeping for a really long time in it temporarily.

DiegoLHendrix reviewed (changes requested):
Looks good. I just got a few grammatical nitpicks and some questions.

mjh9585 commented on .github/workflows/cmake.yml:
```diff
@@ -34,14 +34,9 @@ jobs:
-      - name: F334 Build
-        run: |
-          cmake -DTARGET_DEV=STM32F334x8 -B ${{github.workspace}}/build
+          cmake -DTARGET_DEV=STM32F446xx -DUSE_RTOS=1 -B ${{github.workspace}}/build
```
You might want to add a second option to your CMakeList to enable `USE_RTOS` by default like on line 14 for setting the default target device. This way a more general build check can be written for GitHub actions like in https://github.com/RIT-EVT/project-template/pull/8.

mjh9585 replied:
Also update your runs-on: to `ubuntu-latest` and change `sudo apt-get install clang-format-12` to `sudo apt-get install clang-format-15`. Finally copy the .clang-format from [F4 support and project build checks](https://github.com/RIT-EVT/project-template/pull/8)

aclowmclaughlin replied:
Done!

aclowmclaughlin committed:
Fixed MCuC Imports, made a better description for the class itself, and cleaned up constructor in CPP.

aclowmclaughlin committed:
Made Hardmon Description better.

aclowmclaughlin committed:
Switched std::memcpy references to just memcpy (don't know why the namespacing is not necessary, but it builds now and it wasn't before).

aclowmclaughlin committed:
typo fixes, removes some TODOs that have been completed.

aclowmclaughlin committed:
Typos fixes.

aclowmclaughlin committed:
Added todo explaining Hardmon dictionary.

mjmagee991 committed:
Updated pinout for VCU V2. Commented out unused pins, and added new ones.

aclowmclaughlin committed:
Added (Temporary) new pin array for testing. Will fix later

aclowmclaughlin committed:
commented out incorrect statements

Commit:
Applied Formatting Changes During GitHub Build

mjh9585 commented on include/Hardmon.hpp:
```diff
@@ -11,14 +11,24 @@
 #include <models/Hardmon_Model.hpp>
 
+#include <core/rtos/Initializable.hpp>
+#include <core/rtos/Mutex.hpp>
```
Place these includes with the rest of the core includes and separate out the PowertrainCAN.hpp and models/Hardmon_Model.hpp to its own group after the core imports. In general imports should be grouped alphabetically as Cpp libraries, then EVT-core libraries, then project headers.

mjh9585 commented on include/Hardmon.hpp:
```diff
@@ -142,13 +164,16 @@ class Hardmon : public CANDevice {
      * @return the pointer to the queue
      */
-    core::types::FixedQueue<POWERTRAIN_QUEUE_SIZE, io::CANMessage>* getPowertrainQueue();
+    rtos::Queue* getPowertrainQueue();
```
If possible, you should a make a separate method for reading or writing to the queue externally rather then expose an internal member.

aclowmclaughlin replied:
Completely correct. Did it this way cause I was lazy and it was "temporary" (lol). Fixed.

mjh9585 commented on include/PowertrainCAN.hpp:
```diff
@@ -3,41 +3,59 @@
      * Queue to store CAN messages.
      */
-    core::types::FixedQueue<POWERTRAIN_QUEUE_SIZE, IO::CANMessage> queue;
+    rtos::Queue queue;
```
This should probably be private.

aclowmclaughlin replied:
PowertrainCAN is contained entirely within either MCuC or Hardmon, it's mostly just an easy way for me to keep all of those sorts of relevant messages together. Thus, the queue being public is probably better cause it's only access within the mcuc or hardmon, and PowertrainCAN is private to those classes.

Explaining this made me realize this is probably not the best way to do it- do you think that both classes should privately _inherit_ from PowertrainCAN instead? That's probably the correct object-oriented way to do it.

mjh9585 replied:
I am not sure if inheritance is the right answer here since MCuC and Hardmon both use the accessory and powertrain busses. I was thinking some sort of access method would be better then inheritance. Then you can add any additional semephors before accessing the queue or preform any CAN state updates such as recording message time.

aclowmclaughlin replied:
Bear in mind that MCuC and Hardmon are mutually exclusive in the sense that the target for MCuC and the target for Hardmon are both uploaded to entirely different chips.

mjh9585 commented on include/MCuC.hpp:
```diff
@@ -7,100 +7,116 @@
 #include <core/io/types/CANMessage.hpp>
-#include <core/utils/types/FixedQueue.hpp>
 #include <models/MCuC_Model.hpp>
 
```
Order your imports. Refer to other comment.

mjh9585 commented on include/PowertrainCAN.hpp:
```diff
@@ -3,41 +3,59 @@
+#include <core/rtos/Queue.hpp>
 #include <core/utils/types/FixedQueue.hpp>
 #include <cstddef>
 #include <cstdint>
```
Order!

mjh9585 commented on include/MCuC.hpp:
```diff
@@ -127,42 +140,88 @@ class MCuC : public CANDevice {
      * @return pointer to the fixed queue.
      */
-    core::types::FixedQueue<POWERTRAIN_QUEUE_SIZE, IO::CANMessage>* getPowertrainQueue();
+    rtos::Queue* getPowertrainQueue();
```
Do you need to expose this internal member?

aclowmclaughlin replied:
Fixed.

mjh9585 commented on include/MCuC.hpp:
```diff
@@ -127,42 +140,88 @@ class MCuC : public CANDevice {
+     * reads estop and ignition and then sets the
+     * LVSS signal to 63 (all on) or 0 (all off)
+     */
+    void imagineNeuteredProcess();
```
This should probably be removed at this point.

mjh9585 commented on include/MCuC.hpp:
```diff
@@ -171,7 +230,21 @@ class MCuC : public CANDevice {
+    /**
+     * Unsafe (non-mutexed) Buffer Data that comes in or is sent out over Accessory CAN.
+     */
+    AccessoryCanData_t accessoryCanDataUnsafeBuffer;
```
This feels like there should be a method that takes in a bunch of data to populate the buffer rather then allowing direct access to the structure. Also then all the code for populating this member is self contained in the class and not scattered around elsewhere.

aclowmclaughlin replied:
The issue is that the buffer is not populated all at once, but piecewise.
We would have to have a method for each member of the buffer, which would make adding or removing members (as the model changes what inputs it needs) unwieldy.
In this situation, because the buffer is already considered unsafe, I feel that the raw access to it is probably okay.

mjh9585 commented on include/MCuC.hpp:
```diff
@@ -211,30 +281,38 @@ class MCuC : public CANDevice {
-    static constexpr uint8_t NODE_ID = 255;//TODO: CANopen set node ID
+    static constexpr uint8_t NODE_ID = 0;
+    static constexpr uint8_t LVSS_NODE_ID = 1;
+    static constexpr uint8_t IMU_NODE_ID = 9;
```
Does the Hardmon also need this data? Are there any variables or constants that both need? If so, you should probably make a higher level VCU class that both inherent from to reduce repetitiveness, same is true for any common methods.

aclowmclaughlin replied:
That's a good idea. I will look into this for the next set of updates to VCU.

mjh9585 commented on include/MCuC.hpp:
```diff
@@ -244,11 +322,79 @@ class MCuC : public CANDevice {
+            .Key = CO_KEY(0x1A00 + 0x00, 0x01, CO_OBJ_D___R_),
+            .Type = CO_TUNSIGNED32,
+            .Data = (CO_DATA) CO_LINK(0x2200 + 0x00, 0x00 + 0x01, PDO_MAPPING_UNSIGNED16),
+        },
```
Instead of writing out these keys manually, make a corrected macro to override the default one. Example in https://github.com/RIT-EVT/REV3-TMS/pull/1

mjh9585 commented on include/MCuC.hpp:
```diff
@@ -244,11 +322,79 @@ class MCuC : public CANDevice {
+            .Key = CO_KEY(0x2200 + 0x00, 0x01, CO_OBJ____PRW),
+            .Type = CO_TUNSIGNED16,
+            .Data = (CO_DATA) &accessoryCanDataUnsafeBuffer.LVSS_out_EnableBoardSignal,
+        },
```
This can use the normal DATA_LINK macros, just start the id with 0x100.

mjh9585 commented on include/PowertrainCAN.hpp:
```diff
@@ -3,41 +3,59 @@
      */
     enum MessageIDs {
+        // motor controller receive message ids
+        MC_TEMPS_1_ID = (uint32_t) 0xA0,
```
Why do you need the cast here? If you want to make sure the enum is a uint32 then you should define the enum as `enum MessageIDs : uint32_t`.

aclowmclaughlin replied:
ngl didn't know you could do that. Figured that this would force them to be uint32s. Thanks! Fixed!

mjh9585 commented on targets/REV3-MCUC/main.cpp:
```diff
@@ -96,71 +253,202 @@ int main() {
+ *
+ * @param args the arguments for this thread
+ */
+[[noreturn]] void modelThreadEntry(modelThreadArgs_t* args) {
```
Is it possible to put all of these thread entries in the associated cpp file as a global static method to clean up main? Also since the MCuC is initalizable, can it hold all of the thread variables and call init on them?

aclowmclaughlin replied:
Both are possible. I think that generally the MCuC holding the threads is inadvisable because to me, threads are similar to GPIOs in that they should be explicitly initialized in the main target. 

I'm a little confused about what you mean by the "associated cpp file".

mjh9585 commented on targets/REV3-MCUC/main.cpp:
```diff
@@ -96,71 +253,202 @@ int main() {
+        //suspends if there are no messages to receive
+        queue->receive(&message, rtos::TXWait::TXW_WAIT_FOREVER);
+        args->mcuc->handlePowertrainCanMessage(message);
+    }
```
Also instead of having to access internal members externally to do this, could you move the while loop to an class method that the entry function just calls? This would significantly simplify this function and remove the need for the getPowertrainQueue method.

aclowmclaughlin replied:
I changed it so that it now calls mcuc->recieveFromPowertrainQueue() then calls handlePowerTrainCANMessage(). 
I think that having them be two methods gives us room to do stuff with the message in between (not relevant now but could be eventually). 

As to having the entire loop be a method from the MCuC, I think that it will make it less obvious what the thread is doing and that the thread is infinite looping, which would be generally bad, I think.

mjh9585 replied:
But if MCuC is supposed to do something else with the message then that should be handled in MCuC. Also MCuC is your top level class that directs all of the processing so I would argue that if something needs to be done to the message, MCuC should be in control. 

Also putting the infinite loop in the class will help with needing access to internal members. The problem of it being less obvious is a method naming and documentation issue not a structural one. Additionally, moving the entry function into the associated Cpp file will help with the ambiguity of what is happening since the infinite loop method could be placed right below it and be in scope of the class.

mjh9585 commented on include/PowertrainCAN.hpp:
```diff
@@ -75,7 +93,7 @@ class PowertrainCAN {
      * @return whether or not forward enable is on.
      */
-    bool parseHIBForwardEnable(IO::CANMessage& message);
+    bool parseHIBForwardEnable(io::CANMessage& message);
```
Since there are a lot of boards that the VCU has to talk to, we might want to consider moving all of this parsing to separate board classes that takes in a CAN message from that respective board and extracts the relevant data. Then a set of completed data structs could be passed around to load in the model. Not something supper critcal for this pull but something to consider in for the next.

aclowmclaughlin replied:
That's not a bad idea at all. My thought at the time was that for each board the number of messages is relatively low, so instantiating a large number of classes each with 1 or 2 methods would be relatively redundant, but there also aren't too many boards on the powertrain CAN loop, so it could work out.

mjh9585 replied:
But honestly instantiating more classes isn't really that expensive. If anything it makes it more explicit and clear that the board talks to a list of other ones.

mjh9585 commented on include/PowertrainCAN.hpp:
```diff
@@ -147,19 +167,19 @@ class PowertrainCAN {
     uint8_t HardmonSelfTestResponsePayload = 3;
     ///the hardmon self test message
-    IO::CANMessage HardmonSelfTestResponse = IO::CANMessage(HARDMON_SELF_TEST_MESSAGE_ID, 1, &HardmonSelfTestResponsePayload, false);
+    io::CANMessage HardmonSelfTestResponse = io::CANMessage(HARDMON_SELF_TEST_MESSAGE_ID, 1, &HardmonSelfTestResponsePayload, false);
```
Do these change? If not they can be marked constant.

aclowmclaughlin replied:
good point. only the payloads change.

aclowmclaughlin replied:
had to undo this- the CAN send methods do not maintain const-ness of the pointer, so it doesn't play nice. Better to have them not const the whole time.

mjh9585 commented on src/Hardmon.cpp:
```diff
@@ -1,11 +1,25 @@
+        return status;
+    } else {
+        return powertrainCAN.init(pool);
+    }
```
These members could be placed in an Initializable array and passed back to `bulkInitialize` instead of manually calling the init methods. This would be much safer than having to know how to initialize items correctly and would provide more consistent behavior.

mjh9585 commented on src/MCuC.cpp:
```diff
@@ -1,8 +1,27 @@
+        return status;
+    } else {
+        return powertrainCAN.init(pool);
+    }
```
These members could be placed in an Initializable array and passed back to `bulkInitialize` instead of manually calling the init methods. This would be much safer than having to know how to initialize items correctly and would provide more consistent behavior.

mjh9585 commented on src/PowertrainCAN.cpp:
```diff
@@ -65,4 +64,8 @@ void PowertrainCAN::sendHardmonSelfTestResponse() {
 }
 
+core::rtos::TXError PowertrainCAN::init(rtos::BytePoolBase& pool) {
+    return queue.init(pool);
```
These members could be placed in an Initializable array and passed back to `bulkInitialize` instead of manually calling the init methods. This would be much safer than having to know how to initialize items correctly and would provide more consistent behavior.

aclowmclaughlin replied:
(this one is only one initilizable so I'm gonna leave it as is.

mjh9585 commented on src/Hardmon.cpp:
```diff
@@ -1,11 +1,25 @@
 
-Hardmon::Hardmon(HardmonGPIO gpio, io::CAN& ptCAN) : powertrainCAN(ptCAN), gpios(gpio) {
+Hardmon::Hardmon(HardmonGPIO gpio, io::CAN& ptCAN) : powertrainCAN(ptCAN), gpios(gpio),
+                                                     mutex((char*) "Hardmon Mutex", true),
```
Why do you have to cast the string to a char pointer? Is this necessary? Is this safe? Could this initialization lead to undefined behavior?

aclowmclaughlin replied:
The String is cast to a char* actually because it leads to "safer" behaviour. As far as I'm aware, this is because the c methods that eventually store the string for whatever reason do not mark it as const. In C++ a string literal is a (const char*). If at any point that string literal is used without a const, it is not in C++ standards to implicitly cast it. It still will implicitly cast it, but it gives a clang warning.

So, explicitly casting them prevents that clang warning: 
![image](https://github.com/user-attachments/assets/f2e2d4c9-fa90-4247-ab37-013bccf3735f)
(warning you get when removing the (char*) cast)

It won't matter anyway because the names are actually copied over to a fixed-size name buffer immediately, so the string literal is used only during initialization, and additionally the names are not supposed to be modified. Honestly, it's a pain that ThreadX requires names, but they can be useful for debugging so I left them in.

mjh9585 commented on src/Hardmon.cpp:
```diff
@@ -21,7 +35,9 @@ uint8_t Hardmon::getNodeID() {
 void Hardmon::handlePowertrainCanMessage(io::CANMessage& message) {
     switch (message.getId()) {
     case dev::PowertrainCAN::HIB_MESSAGE_ID:
+        mutex.get(rtos::TXWait::TXW_WAIT_FOREVER);
         forwardEnable = powertrainCAN.parseHIBForwardEnable(message);
+        mutex.put();
```
I am assuming that hardmon is not ready for the rest of the hib messages and will be handle ina future pull. Right?

aclowmclaughlin replied:
Yes. Hardmon is especially light on implementation details- I focused mostly on MCuC.

mjh9585 commented on src/Hardmon.cpp:
```diff
@@ -32,18 +48,26 @@ void Hardmon::handlePowertrainCanMessage(io::CANMessage& message) {
+    mutex.get(rtos::TXWait::TXW_WAIT_FOREVER);
+    accessoryCanDataSafeBuffer.LVSS_in_HVCurrent = accessoryCanDataUnsafeBuffer.LVSS_in_HVCurrent;
+    accessoryCanDataSafeBuffer.LVSS_in_PowerSwitchCurrents = accessoryCanDataUnsafeBuffer.LVSS_in_PowerSwitchCurrents;
+    accessoryCanDataSafeBuffer.LVSS_in_PowerSwitchErrorStatus = accessoryCanDataUnsafeBuffer.LVSS_in_PowerSwitchErrorStatus;
```
Could this just be a memcpy rather than a list of assignments? If you add a special marker field in the struct you could calculate the size of the copy, making it more flexible for future additions. A union with a set of arrays would also work.

aclowmclaughlin replied:
Yep, fixed it. (It was fixed in MCuC and I forgot to copy it over).
In that note- I may create a shared AccessoryCANOpen Class just like I did the powertrainCAN class in order to not have these things have to be duplicated between classes.

mjh9585 commented on src/MCuC.cpp:
```diff
@@ -17,116 +36,200 @@ uint8_t MCuC::getNodeID() {
     case dev::PowertrainCAN::HARDMON_SELF_TEST_MESSAGE_ID:
+        mutex.get(rtos::TXWait::TXW_WAIT_FOREVER);
         powertrainCANSelfTestIn = true;
+        mutex.put();
```
Since you are always getting and putting the mutex at the beginning and end of each case you should just wrap the switch in the mutex to reduce redundant code.

mjh9585 commented on src/MCuC.cpp:
```diff
@@ -17,116 +36,200 @@ uint8_t MCuC::getNodeID() {
 
+void MCuC::sendOutputDataToUnsafeBuffer() {
+    mutex.get(rtos::TXWait::TXW_WAIT_FOREVER);
+    accessoryCanDataUnsafeBuffer.LVSS_out_EnableBoardSignal = accessoryCanDataSafeBuffer.LVSS_out_EnableBoardSignal;
```
I would still do a memcpy here just to make this more flexible for future modifications. Also if you don't want to copy everything for speed, you can put an additional marker in the struct to separate the incoming and outgoing signals. then you can group the signals you want to copy in a given direction. Also a union with a set of arrays would work.

aclowmclaughlin replied:
Agree. Used union with set of arrays.

mjh9585 commented on include/Hardmon.hpp:
```diff
@@ -157,7 +182,26 @@ class Hardmon : public CANDevice {
+
+    void sendOutputDataToUnsafeBuffer();
+
+    void sendInputDataToSafeBuffer();
```
If this process is only controlled by the class then these should be made private.

mjh9585 commented on src/MCuC.cpp:
```diff
@@ -17,116 +36,200 @@ uint8_t MCuC::getNodeID() {
-    lvssOn = gpios.lvssStatusGPIO.readPin() == IO::GPIO::State::HIGH;
-    mcOn = gpios.mcStatusGPIO.readPin() == IO::GPIO::State::HIGH;
+    //lvssOn = gpios.lvssStatusGPIO.readPin() == io::GPIO::State::HIGH;
+    mcOn = gpios.mcStatusGPIO.readPin() == io::GPIO::State::HIGH;
```
Does this need to be in the mutex?

mjh9585 commented on src/MCuC.cpp:
```diff
@@ -17,116 +36,200 @@ uint8_t MCuC::getNodeID() {
+    }
+    sendOutputDataToUnsafeBuffer();
+    mutex.put();
 }
```
Remove?

mjh9585 commented on targets/REV3-MCUC/main.cpp:
```diff
@@ -63,11 +159,67 @@ int main() {
+    //    rtos::tsio::ThreadUART threadUART(uart, UART_THREAD_STACK_SIZE,
+    //                                      UART_THREAD_PRIORITY,
+    //                                      UART_THREAD_PREEMPT_THRESHOLD,
+    //                                      UART_THREAD_TIME_SLICE);
```
Do you still need this?

aclowmclaughlin replied:
Technically all the logging is being done with an unsafe uart instance, but in this situation that's probably the lesser of two evils- nothing else uses uart, so it's still honestly generally quite threadsafe. Ironically the thread safe uart, though it is _threadsafe_, is actually probably more of a danger in this situation- simply having another thread handling uart is more likely to modify behaviour than using the unthreaded uart (that, or the uart thread is so low priority that it basically never runs).

TLDR; nah, probably not. Removed it.

mjh9585 commented on targets/REV3-MCUC/main.cpp:
```diff
@@ -63,11 +159,67 @@ int main() {
+            io::getGPIO<vcu::MCuC::ESTOP_B_PIN>(io::GPIO::Direction::OUTPUT),
+            io::getGPIO<vcu::MCuC::IGNITION_B_PIN>(io::GPIO::Direction::OUTPUT),
+            io::getGPIO<vcu::MCuC::INTERLOCK_PIN>(io::GPIO::Direction::OUTPUT),
+        }};
```
Wow is all I have to say. Very clean, no BS. No further comment...

mjh9585 replied:
Ok, is the TODO for this pull or the next?

aclowmclaughlin replied:
Next- literally like on Saturday Blake called me and asked for me to make some changes so they could test the new VCU board.

mjh9585 commented on targets/REV3-MCUC/main.cpp:
```diff
@@ -96,71 +253,202 @@ int main() {
+    // Initialize Threads
+
+    /// eventflag that triggers the model to run
+    rtos::EventFlags modelTriggerFlag((char*) "Model Trigger Flag");
```
Could these flags also be used for other things to signify the model state to the health thread such as a time overrun event, the model getting stale data, or even just a model complete flag to kick off other actions?

aclowmclaughlin replied:
Yeah they totally could. For now it's just the model trigger, but it is certainly possible to use the other flags in the group for other parts. If we were to do that, I would rename it to be more descriptive.

mjh9585 commented on targets/REV3-Hardmon/main.cpp:
```diff
@@ -37,87 +131,45 @@ namespace time = core::time;
+    auto* queue = (rtos::Queue*) priv;
+    if (queue != nullptr) {
+        //TODO: determine if WaitForever is what we want to do in the interrupt- could be bad
+        queue->send(static_cast<void*>(&message), rtos::TXWait::TXW_WAIT_FOREVER);
```
You do not want to wait forever although I don't think it matters since you are in interrupt context and I believe all methods return immediately.

mjh9585 replied:
Also here in Hardmon you write directly to the Queue but in MCuC you write to the normal queue is there a reason for the difference?

mjh9585 replied:
Also also, you pass in a fixed queue but interpret it as a rtos queue.

aclowmclaughlin replied:
Yeah the hardmon version was a little messed up. AccessoryCAN should be a fixedQueue. 
I don't know what happens if a thread gets slept in the middle of an interrupt-- might be worth looking into.

mjh9585 commented on targets/REV3-MCUC/main.cpp:
```diff
@@ -39,19 +135,19 @@ namespace time = core::time;
-void powertrainCANInterrupt(IO::CANMessage& message, void* priv) {
-    auto* queue = (core::types::FixedQueue<POWERTRAIN_QUEUE_SIZE, IO::CANMessage>*) priv;
+void powertrainCANInterrupt(io::CANMessage& message, void* priv) {
+    auto* queue = (core::types::FixedQueue<POWERTRAIN_QUEUE_SIZE, io::CANMessage>*) priv;
```
You are passing in an rtos queue, but interpreting it as a fixed queue.

aclowmclaughlin replied:
Yeah fixed this.

mjh9585 reviewed (changes requested):
Lots of nitpicky things but overall nice work. Have fun with this and I am sorry in advance. :)

aclowmclaughlin committed:
Changed cmake settings, updated clang format and github actions

aclowmclaughlin committed:
Merge remote-tracking branch 'origin/feature/aclowmclaughlin/VCU_RTOS' into feature/aclowmclaughlin/VCU_RTOS

aclowmclaughlin committed:
fixed clang format

Commit:
Applied Formatting Changes During GitHub Build

DiegoLHendrix commented on include/Hardmon.hpp:
```diff
@@ -11,14 +11,24 @@
 /**
- * Driver for the Hardware Monitor
+ * The Hardmon (Hardware Monitor) is one of two microcontrollers on the VCU board.
+ * It monitors the MCuC in order to determine whether or not it is operating safely and logically.
```
Since this is the first time in this file the MCuC is mentioned

```suggestion
It monitors the MCuC (Motor Controller Microcontroller) in order to determine whether or not it is operating safely and logically.```

aclowmclaughlin committed:
Changed both MCuC and Hardmon to have a recieveFromPowertrainQueue() and sendToPowertrainQueue() methods instead of just a getPowertrainQueue() method

changed the targets (mains) of both to work with the new methods.

Commit:
Applied Formatting Changes During GitHub Build

aclowmclaughlin committed:
changed include in MCuC and removed unnecessary include in Hardmon.

aclowmclaughlin committed:
Merge remote-tracking branch 'origin/feature/aclowmclaughlin/VCU_RTOS' into feature/aclowmclaughlin/VCU_RTOS

aclowmclaughlin committed:
reordered PowertrainCAN includes and removed deprecated FixedQueue include.

aclowmclaughlin committed:
misc PowertrainCAN changes from Heller's PR feedback

aclowmclaughlin committed:
reverted const changes to PowertrainCAN messages. didn't play well with CAN library (the const gets dropped when sent)

Commit:
Applied Formatting Changes During GitHub Build

aclowmclaughlin committed:
Modified MCuC and Hardmon init() method to use bulk initialize instead of initializing them one by one.

Also copied updated buffers for accessoryCAN from MCuC and Hardmon and changed corresponding methods.

Commit:
Applied Formatting Changes During GitHub Build

aclowmclaughlin committed:
Switched mutex gets in handlePowertrainCANMessage methods to before and after the switch statement.

Fixed typo in receiveFromPowertrainQueue (receive was spelled recieve)

Switched the accessoryCanData_t to a union containing two structs in order to make memcpying separate parts of them more viable (then switched to using memcpy for both sendInputDataToSafeBuffer and sendOutputDataToUnsafeBuffer.

Commit:
Applied Formatting Changes During GitHub Build

aclowmclaughlin committed:
cleaned up commented out stuff.

DiegoLHendrix reviewed (approved):
Good work

aclowmclaughlin committed:
Diego Change.

Co-authored-by: Diego <119289719+DiegoLHendrix@users.noreply.github.com>

Commit:
Applied Formatting Changes During GitHub Build

tmb5932 committed:
updated pins

tmb5932 committed:
added todo's for documenting purposes

tmb5932 committed:
updated naming and order of I/O pins

Commit:
Applied Formatting Changes During GitHub Build

tmb5932 reviewed (approved):
Ok all I found that I was concerned (besides prev mentioned things) about are things ive already altered in my local version, so i think this is good.

I copied all the concerns/ideas that people (mainly heller) gave, so they are now part of my VCU todo list
